### ISSUE TITLE
Table queries stopped working after upgrade to latest 3.21.0 version 

### ISSUE THREAD

**Issue by BennyM** [Azure/Azurite#1787]
### Which service(blob, file, queue, table) does this issue concern?
Table

### Which version of the Azurite was used?
3.21.0

### Where do you get Azurite? (npm, DockerHub, NuGet, Visual Studio Code Extension)
npm

### What's the Node.js version?
v 16.18.0

### What problem was encountered?
Bad requests when query gets executed. Worked in the previous version.

### Steps to reproduce the issue?

```
2023-01-10T10:52:25.437Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 info: DispatchMiddleware: Operation=Table_QueryEntities
2023-01-10T10:52:25.437Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 verbose: AuthenticationMiddlewareFactory:createAuthenticationMiddleware() Validating authentications.
2023-01-10T10:52:25.437Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 info: TableSharedKeyLiteAuthenticator:validate() Start validation against account shared key authentication.
2023-01-10T10:52:25.438Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 info: TableSharedKeyLiteAuthenticator:validate() [STRING TO SIGN]:"Tue, 10 Jan 2023 10:52:25 GMT\n/devstoreaccount1/devstoreaccount1/customers()"
2023-01-10T10:52:25.438Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 info: TableSharedKeyLiteAuthenticator:validate() Calculated authentication header based on key1: SharedKeyLite devstoreaccount1:gik6OwVuxEbxe51oXuBSWckmDpSpLpciLNkzZrOuaG8=
2023-01-10T10:52:25.438Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 info: TableSharedKeyLiteAuthenticator:validate() Signature 1 matched.
2023-01-10T10:52:25.438Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 verbose: DeserializerMiddleware: Start deserializing...
2023-01-10T10:52:25.438Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 info: HandlerMiddleware: DeserializedParameters={"options":{"queryOptions":{"format":"application/json;odata=minimalmetadata","filter":"(true and RowKey eq 'settings') and ((false or PartitionKey eq '1P23QNX4') or PartitionKey eq 'ABC_123') or PartitionKey eq '1A2BC3DE'"},"requestId":"01f67662-0ec0-4413-a45b-d38de35d90a6","dataServiceVersion":"3.0"},"version":"2020-12-06"}
2023-01-10T10:52:25.443Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 error: ErrorMiddleware: Received a MiddlewareError, fill error information to HTTP response
2023-01-10T10:52:25.444Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 error: ErrorMiddleware: ErrorName=StorageError ErrorMessage="The query condition specified in the request is invalid."  ErrorHTTPStatusCode=400 ErrorHTTPStatusMessage=undefined ErrorHTTPHeaders={"x-ms-error-code":"InvalidInput","x-ms-request-id":"2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0","x-ms-version":"2021-10-04"} ErrorHTTPBody="{\"odata.error\":{\"code\":\"InvalidInput\",\"message\":{\"lang\":\"en-US\",\"value\":\"The query condition specified in the request is invalid.\\nRequestId:2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0\\nTime:2023-01-10T10:52:25.443Z\"}}}" ErrorStack="\"StorageError: The query condition specified in the request is invalid.\\n    at Function.getQueryConditionInvalid (/usr/local/lib/node_modules/azurite/dist/src/table/errors/StorageErrorFactory.js:79:16)\\n    at Function.generateQueryForPersistenceLayer (/usr/local/lib/node_modules/azurite/dist/src/table/persistence/LokiTableStoreQueryGenerator.js:27:49)\\n    at LokiTableMetadataStore.queryTableEntities (/usr/local/lib/node_modules/azurite/dist/src/table/persistence/LokiTableMetadataStore.js:252:67)\\n    at TableHandler.queryEntities (/usr/local/lib/node_modules/azurite/dist/src/table/handlers/TableHandler.js:316:81)\\n    at /usr/local/lib/node_modules/azurite/dist/src/table/generated/middleware/HandlerMiddlewareFactory.js:58:18\\n    at /usr/local/lib/node_modules/azurite/dist/src/table/generated/ExpressMiddlewareFactory.js:77:63\\n    at Layer.handle [as handle_request] (/usr/local/lib/node_modules/azurite/node_modules/express/lib/router/layer.js:95:5)\\n    at trim_prefix (/usr/local/lib/node_modules/azurite/node_modules/express/lib/router/index.js:328:13)\\n    at /usr/local/lib/node_modules/azurite/node_modules/express/lib/router/index.js:286:9\\n    at Function.process_params (/usr/local/lib/node_modules/azurite/node_modules/express/lib/router/index.js:346:12)\""
2023-01-10T10:52:25.444Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 error: ErrorMiddleware: Set HTTP code: 400
2023-01-10T10:52:25.444Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 error: ErrorMiddleware: Set HTTP Header: x-ms-error-code=InvalidInput
2023-01-10T10:52:25.445Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 error: ErrorMiddleware: Set HTTP Header: x-ms-request-id=2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0
2023-01-10T10:52:25.445Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 error: ErrorMiddleware: Set HTTP Header: x-ms-version=2021-10-04
2023-01-10T10:52:25.445Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 error: ErrorMiddleware: Set content type: application/json;odata=minimalmetadata;streaming=true;charset=utf-8
2023-01-10T10:52:25.445Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 error: ErrorMiddleware: Set HTTP body: "{\"odata.error\":{\"code\":\"InvalidInput\",\"message\":{\"lang\":\"en-US\",\"value\":\"The query condition specified in the request is invalid.\\nRequestId:2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0\\nTime:2023-01-10T10:52:25.443Z\"}}}"
2023-01-10T10:52:25.453Z 2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0 info: EndMiddleware: End response. TotalTimeInMS=18 StatusCode=400 StatusMessage=Bad Request Headers={"server":"Azurite-Table/3.21.0","x-ms-error-code":"InvalidInput","x-ms-request-id":"2f1bb080-5d32-4fa4-bdf4-fdd387c89bb0","x-ms-version":"2021-10-04","content-type":"application/json;odata=minimalmetadata;streaming=true;charset=utf-8"}

```

### Have you found a mitigation/solution?
Use previous version

**Comment by blueww** [comment]
@BennyM 
Thanks for raising this issue!
We will look deeper into it and update you.

@edwin-huber 
It looks related with with the true/false in query `(true and RowKey eq 'settings') and ((false or PartitionKey eq '1P23QNX4') or PartitionKey eq 'ABC_123') or PartitionKey eq '1A2BC3DE'`, and might be related with the change in https://github.com/Azure/Azurite/pull/1521. Not sure if you have any idea on this？

**Comment by blueww** [comment]
It looks the error is in function: https://github.com/Azure/Azurite/blob/e383fc08c64d3f59c089dc7e338a86ec300c636a/src/table/persistence/QueryTranscriber/StatePredicateFinished.ts#L343

For checkSingleTerm() function, currently the code assume it should only be ParensOpen/ParensClose/Operator, but actually it can also be a boolean value. 
I will raise a fix PR soon.

**Comment by BennyM** [comment]
@blueww to give you some more context. Queries are built with an expression and then given to QueryAsync.

``` 
   var query = PredicateBuilder.False<MyEntity>();
   foreach (var partitionKey in partitionKeys)
   {
      query = query.Or(entity => entity.PartitionKey == partitionKey);
   }
```

Similar to an approach illustrated [here](https://www.albahari.com/nutshell/predicatebuilder.aspx).

**Comment by blueww** [comment]
I have raised a PR to fix this issue.
@edwin-huber, could you please help to review the PR as it change the code raised by you on https://github.com/Azure/Azurite/pull/1521.

**Comment by edwin-huber** [comment]
This totally makes sense... I did not include that case.
I'm reviewing the PR.

**Comment by blueww** [comment]
Close this issue as the fix is already released in Azurite 3.22.0.
